### PR TITLE
[Enhancement] Create factory for histogram and line chart, add brush handle to range brush

### PR DESCRIPTION
**Pull request by heshan0131**
improve histogram brush interaction
Signed-off-by: Shan He <heshan0131@gmail.com>

### CODE REVIEW

File: src/components/common/histogram-plot.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  .histogram-bars {
+    rect {
+      fill: ${props => props.theme.histogramFillOutRange};
+    }
+    rect.in-range {
+      fill: ${props => props.theme.histogramFillInRange};
+    }
+  }
+`;
+function HistogramPlotFactory() {
```
Comment (macrigiuseppe) [inline src/components/common/histogram-plot.js:23]: nit: missing new line

File: src/components/common/histogram-plot.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  const HistogramPlot = ({width, height, margin, isRanged, histogram, value, brushComponent}) => {
+    const domain = useMemo(() => [histogram[0].x0, histogram[histogram.length - 1].x1], [
+      histogram
+    ]);
+    const dataId = Object.keys(histogram[0]).filter(k => k !== 'x0' && k !== 'x1')[0];
```
Comment (macrigiuseppe) [inline src/components/common/histogram-plot.js:28]: you can use useMemo here too

File: src/components/common/line-chart.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    stroke: ${props => props.theme.histogramFillOutRange};
+    stroke-dasharray: 1px 4px;
+  }
+
+  .rv-xy-plot__axis__tick__text {
+    font-size: 9px;
+    fill: ${props => props.theme.textColor};
+  }
+`;
+const StyledHint = styled.div`
```
Comment (macrigiuseppe) [inline src/components/common/line-chart.js:55]: empty line missing

File: src/components/common/line-chart.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  font-size: 9px;
+  margin: 4px;
+  padding: 3px 6px;
+  pointer-events: none;
+  user-select: none;
+`;
+
+const HintContent = ({x, y, format}) => (
+  <StyledHint>
+    <div className="hint--x">{format(x)}</div>
```
Comment (macrigiuseppe) [inline src/components/common/line-chart.js:68]: could you pass the result of format(x) as input of this component?

File: src/components/common/line-chart.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    <div className="row">{y}</div>
+  </StyledHint>
+);
```
Comment (macrigiuseppe) [inline src/components/common/line-chart.js:71]: new empty line

File: src/components/common/line-chart.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+const MARGIN = {top: 0, bottom: 0, left: 0, right: 0};
+function LineChartFactory() {
```
Comment (macrigiuseppe) [inline src/components/common/line-chart.js:73]: new empty line

File: src/components/common/range-brush.js [hunk 7]
```diff
@@ -43,6 +43,21 @@ function moveRight(startSel, selection) {
...
+// style brush resize handle
+// https://github.com/crossfilter/crossfilter/blob/gh-pages/index.html#L466
+const getHandlePath = props => {
+  return function brushResizePath(d) {
+    const e = Number(d.type === 'e');
+    const x = e ? 1 : -1;
+    const h = 39;
+    const w = 4.5;
+    const y = (props.height - h) / 2;
+    // return `M0,57.3c2.5,0,4.5,2,4.5,4.5v29.3c0,2.5-2,4.5-4.5,4.5V57.3z`
```
Comment (macrigiuseppe) [inline src/components/common/range-brush.js:55]: remove commented lines

File: src/components/common/range-plot.js [hunk 8]
```diff
@@ -18,201 +18,118 @@
...
+  margin-bottom: ${props => props.theme.sliderBarHeight}px;
+  display: flex;
+  position: 'relative';
+`;
 
-    domainSelector = props => props.lineChart && props.lineChart.xDomain;
-    hintFormatter = createSelector(this.domainSelector, domain =>
-      getTimeWidgetHintFormatter(domain)
+RangePlotFactory.deps = [RangeBrushFactory, HistogramPlotFactory, LineChartFactory];
+export default function RangePlotFactory(RangeBrush, HistogramPlot, LineChart) {
```
Comment (macrigiuseppe) [inline src/components/common/range-plot.js:35]: empty line

File: test/browser/components/common/range-plot.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    plotType: 'histogram',
+    range: [1421315219000, 1421348744000],
+    value: [1421315219000, 1421348744000],
+    width: 137
+  };
+  t.doesNotThrow(() => {
+    const wrapper = mountWithTheme(<RangePlot {...props} />);
+  }, 'Show not fail without histogram');
+
+  // cant test D3 in jsDom for now
```
Comment (macrigiuseppe) [inline test/browser/components/common/range-plot.js:45]: delete commented lines.

is testing only for render failures enough for this component?

**Review by macrigiuseppe — APPROVED** [review]
just few comments

File: test/browser/components/common/range-plot.js (hunk 9 above)
Comment (heshan0131) [reply test/browser/components/common/range-plot.js:45]: for now yea, better than nothing, we can't test d3 with jsDom at all, unless we created a bunch of utils for that component and only testing utils

File: src/components/common/histogram-plot.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    }
+  }
+`;
+
+function HistogramPlotFactory() {
+  const HistogramPlot = ({width, height, margin, isRanged, histogram, value, brushComponent}) => {
+    const domain = useMemo(() => [histogram[0].x0, histogram[histogram.length - 1].x1], [
+      histogram
+    ]);
+    const dataId = Object.keys(histogram[0]).filter(k => k !== 'x0' && k !== 'x1')[0];
```
Comment (heshan0131) [inline src/components/common/histogram-plot.js:29]: it's very cheap operation, and the return value is a string, it's probably more expensive to memorize it than just call it

**Pull request merged**